src/main/java/core/basesyntax/model/impl/PurchaseOperation.java

```diff
@@ -0,0 +1,16 @@
+package core.basesyntax.model.impl;
+
+import core.basesyntax.db.Warehouse;
+import core.basesyntax.model.AbstractOperation;
+import core.basesyntax.model.entities.Product;
+
+public class PurchaseOperation<T extends Product> extends AbstractOperation<T> {
+    public PurchaseOperation(Warehouse<T> warehouse) {
+        super(warehouse);
+    }
+
+    @Override
+    public void execute(T product, Integer amount) {
+        warehouse.getStorage().merge(product, amount, (a, b) -> a - b);
```

Comment:
should we throw an exception here if we don't have enough products?

Reply:
validation will handle such cases before operation execution

Reply:
let's move the checks from the validator class to the necessary operations